From 6c3e82f02afb45f08a3e476fcdf0a87135a8c0e6 Mon Sep 17 00:00:00 2001
From: Marius Vlad <marius.vlad@collabora.com>
Date: Fri, 5 Mar 2021 21:44:26 +0200
Subject: [PATCH 6/8] desktop-shell: Embed keyboard focus handle code when
 activating

We shouldn't be constrained by having a keyboard plugged-in, so avoid
activating/de-activating the window/surface in the keyboard focus
handler and embed it straight into activation part. Input focus relates
to having an input device.

Upstream-Status: Pending MR [https://gitlab.freedesktop.org/wayland/weston/-/merge_requests/570]
Signed-off-by: Marius Vlad <marius.vlad@collabora.com>
---
 desktop-shell/shell.c | 33 +++++++++++++++++----------------
 1 file changed, 17 insertions(+), 16 deletions(-)

diff --git a/desktop-shell/shell.c b/desktop-shell/shell.c
index c4669f11..1f11e088 100644
--- a/desktop-shell/shell.c
+++ b/desktop-shell/shell.c
@@ -1885,22 +1885,8 @@ shell_surface_activate(struct shell_surface *shsurf)
 static void
 handle_keyboard_focus(struct wl_listener *listener, void *data)
 {
-	struct weston_keyboard *keyboard = data;
-	struct shell_seat *seat = get_shell_seat(keyboard->seat);
-
-	if (seat->focused_surface) {
-		struct shell_surface *shsurf = get_shell_surface(seat->focused_surface);
-		if (shsurf)
-			shell_surface_deactivate(shsurf);
-	}
-
-	seat->focused_surface = weston_surface_get_main_surface(keyboard->focus);
-
-	if (seat->focused_surface) {
-		struct shell_surface *shsurf = get_shell_surface(seat->focused_surface);
-		if (shsurf)
-			shell_surface_activate(shsurf);
-	}
+	/* kept for future usage, handle_keyboard_focus() is trigged from
+	 * weston_keyboard_set_focus() */
 }
 
 /* The surface will be inserted into the list immediately after the link
@@ -3836,6 +3822,7 @@ activate(struct desktop_shell *shell, struct weston_view *view,
 	struct workspace *ws;
 	struct weston_surface *old_es;
 	struct shell_surface *shsurf, *shsurf_child;
+	struct shell_seat *shseat = get_shell_seat(seat);
 
 	main_surface = weston_surface_get_main_surface(es);
 	shsurf = get_shell_surface(main_surface);
@@ -3855,6 +3842,20 @@ activate(struct desktop_shell *shell, struct weston_view *view,
 
 	weston_view_activate(view, seat, flags);
 
+	if (shseat->focused_surface) {
+		struct shell_surface *current_focus =
+			get_shell_surface(shseat->focused_surface);
+		if (current_focus)
+			shell_surface_deactivate(current_focus);
+	}
+
+	shseat->focused_surface = main_surface;
+
+	if (shseat->focused_surface) {
+		if (shsurf)
+			shell_surface_activate(shsurf);
+	}
+
 	state = ensure_focus_state(shell, seat);
 	if (state == NULL)
 		return;
-- 
2.30.1

